src/main/java/org/openjdk/jextract/impl/NameMangler.java

```diff
@@ -192,9 +200,16 @@ public Void visitTypedef(Declaration.Typedef typedef, Declaration parent) {
             return null;
         }
 
-        // We may potentially generate a class for a typedef. Make sure
-        // class name is unique in the current nesting context.
-        String javaName = curScope.uniqueNestedClassName(typedef.name());
+        String javaName;
+        if (Utils.getAnonymousStructTypedef(typedef) != null) {
+            // a typedef of a struct/union without a name is special-cased,
+            // so that the structs "inherits" the typedef name.
+            javaName = typedef.name();
+        } else {
+            // We may potentially generate a class for a typedef. Make sure
+            // class name is unique in the current nesting context.
+            javaName = curScope.uniqueNestedClassName(typedef.name());
```

Comment:
This logic is a bit subtle: we have to avoid calling `uniqueNestedClassName` in case there's a struct that will inherit the typedef name - otherwise the mangler will see _two_ attempts to add the typedef name to the current scope, which will result in extra mangling.

Reply:
This seems a bit ad-hoc... Suggestion: the `Scope` constructor tries to generate a unique nested class name if `parent` is not `null`. That's only the case when calling `newStruct`, which is only called from `visitScoped`. So, I think we can move the logic that computes the nested class name from the `Scope` constructor to `visitScoped` (to the `else` branch this patch adds). Then we can just pass the unmangled name from the `if` branch to `newStruct`, and there should be no need to have this logic here.

Reply:
Thanks for the suggestion - I did think of something similar, but had not realized we only needed the mangling in one place. Updated the code, looks a bit better I think.